fdb-record-layer-core/src/main/java/com/apple/foundationdb/record/query/plan/plans/RecordQuerySelectorPlan.java

```diff
@@ -0,0 +1,343 @@
+        if (!continuation.isEmpty()) {
+            // Continuation overrides the lambda selection.
+            return (int)continuation.getSelectedPlanIndex();
+        } else {
+            // Select a plan according to the relative priorities
+            return planSelector.selectPlan(getChildren());
+        }
+    }
+
+    private static class SelectorPlanCursor implements RecordCursor<QueryResult> {
+        // The index of the selected plan within the parent selector plan
+        private final long selectedPlanIndex;
+        // Inner cursor to provide record inflow
+        @Nonnull
+        private final RecordCursor<QueryResult> inner;
+        @Nullable FDBStoreTimer timer;
+
+        public SelectorPlanCursor(final long selectedPlanIndex, @Nonnull final RecordCursor<QueryResult> inner, @Nullable FDBStoreTimer timer) {
+            this.inner = inner;
+            this.selectedPlanIndex = selectedPlanIndex;
+            this.timer = timer;
+        }
+
+        @Nonnull
+        @Override
+        public CompletableFuture<RecordCursorResult<QueryResult>> onNext() {
+            return inner.onNext().thenApply(this::calculateCursorResult);
+        }
+
+        @Override
+        public void close() {
+            inner.close();
+        }
+
+        @Nonnull
+        @Override
+        public Executor getExecutor() {
+            return inner.getExecutor();
+        }
+
+        @Override
+        public boolean accept(@Nonnull RecordCursorVisitor visitor) {
+            if (visitor.visitEnter(this)) {
+                inner.accept(visitor);
+            }
+            return visitor.visitLeave(this);
+        }
+
+        private RecordCursorResult<QueryResult> calculateCursorResult(final RecordCursorResult<QueryResult> innerResult) {
+            final long startTime = System.nanoTime();
+            if (innerResult.hasNext()) {
+                SelectorContinuation continuation = new SelectorContinuation(selectedPlanIndex, innerResult.getContinuation().toBytes(), false);
+                logTimer(startTime);
+                return RecordCursorResult.withNextValue(innerResult.get(), continuation);
+            } else {
+                // Inner is done - return a no-next result with wrapped continuation and the same no-next reason and isEnd.
+                SelectorContinuation continuation = new SelectorContinuation(selectedPlanIndex, innerResult.getContinuation().toBytes(), innerResult.getContinuation().isEnd());
+                logTimer(startTime);
+                return RecordCursorResult.withoutNextValue(continuation, innerResult.getNoNextReason());
+            }
+        }
+
+        private void logTimer(final long startTime) {
+            if (timer != null) {
+                timer.record(FDBStoreTimer.Events.QUERY_SELECTOR, System.nanoTime() - startTime);
+            }
+        }
+    }
+
+    public interface PlanSelector extends PlanHashable {
+        int selectPlan(List<RecordQueryPlan> plans);
```

Comment:
minor: It's a bit six of one, half dozen of another, but part of me wonders if this should return a `RecordQueryPlan` instead of the position into the list. Just because some implementations (e.g., ones that iterate over the plans before picking the "best" one, based on some criterion) might prefer that interface.

Reply:
True, but that would also force the `PlanSelector` to store the list of plans, which it does not have to do now (since the selection is pure mathematical operation). I'd rather defer this change until it is proven necessary.

Reply:
Couldn't it get it from the `plans` list that are passed in as a parameter?

Reply:
Yes, it could, but then, we also need the index of the plan (so that we can store it in the continuation), and so would have to create a structure that has both plan and index, and the code would get more complex...